pkg/cloud/vsphere/services/govmomi/esxi/clone.go

```diff
@@ -17,12 +17,271 @@ limitations under the License.
+		deviceSpecs = append(deviceSpecs, device)
+	}
+
+	spec.DeviceChange = deviceSpecs
+	spec.Files = &types.VirtualMachineFileInfo{
+		VmPathName: fmt.Sprintf("[%s]", vm.Datastore.Name()),
+	}
+
+	var extraConfig extra.Config
+	extraConfig.SetCloudInitUserData(userData)
+	spec.ExtraConfig = extraConfig
+
+	task, err := CreateVM(vm, spec)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create vm")
+	}
+	ctx.Logger.V(4).Info("Task submitted: " + task.Reference().Value)
+	ctx.MachineStatus.TaskRef = task.Reference().Value
+	return nil
+}
+
+// VmContext is a context that adds govmomi objects to a MachineContext
+type VmContext struct {
+	*context.MachineContext
+	Src        *object.VirtualMachine
+	Folder     *object.Folder
+	Datastore  *object.Datastore
+	Host       *object.HostSystem
+	Datacenter *object.Datacenter
+	Pool       *object.ResourcePool
+}
+
+// CreateVM submits the folder.createVM task
+func CreateVM(ctx *VmContext, spec *types.VirtualMachineConfigSpec) (*object.Task, error) {
+	ctx.Logger.V(6).Info("esxi cloning machine", "clone-spec", spec)
+
+	folders, err := ctx.Datacenter.Folders(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return folders.VmFolder.CreateVM(ctx, *spec, ctx.Pool, ctx.Host)
+}
+
+// NewVmContext creates a new VmContext from a MachineContext
+func NewVmContext(ctx *context.MachineContext) (*VmContext, error) {
+	ctx = context.NewMachineLoggerContext(ctx, "esxi")
+	ctx.Logger.V(6).Info("starting clone process")
+	src, err := template.FindTemplate(ctx, ctx.MachineConfig.MachineSpec.VMTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	host, err := src.HostSystem(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "hostSystem failed")
+	}
+
+	datastore, err := ctx.Session.Finder.DatastoreOrDefault(ctx, ctx.MachineConfig.MachineSpec.Datastore)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get datastore for machine %q", ctx)
+	}
+
+	ctx.Logger.V(2).Info("attempting to use host resource pool")
+	pool, err := host.ResourcePool(ctx.Context)
+	if err != nil {
+		return nil, errors.Wrap(err, "host resource pool failed")
+	}
+
+	dc, err := ctx.Session.Finder.DefaultDatacenter(ctx.Context)
+	if err != nil {
+		return nil, err
+	}
+
+	vmCtx := VmContext{
+		MachineContext: ctx,
+		Datastore:      datastore,
+		Pool:           pool,
+		Datacenter:     dc,
+		Host:           host,
+		Src:            src,
+	}
+	return &vmCtx, nil
+}
+
+// GetVirtualMachineMO is a convenience method that wraps fetching the VirtualMachine
+// MO from its higher-level object.
+func (ctx *VmContext) GetVirtualMachineMO() (*mo.VirtualMachine, error) {
+	var props mo.VirtualMachine
+	if err := ctx.Src.Properties(ctx.Context, ctx.Src.Reference(), nil, &props); err != nil {
+		return nil, err
+	}
+	return &props, nil
+}
+
+// GetNics creates and returns network devices from those specifid under Network.Devices
+func GetNics(ctx *VmContext) ([]types.BaseVirtualDeviceConfigSpec, error) {
```

Comment:
Hi @moshloop,

`s/GetNIcs/getNetworkSpecs` to be consistent with the vCenter package?